Problems when using Octant with multiple kubeconfig

Hello

This is similar to #147 but I don't think this is resolved
I have multiple kubeconfig files that I'm using through the **KUBECONFIG** env var

**What I expected?**

- By default, running octant with no params will auto select the active context from the merged KUBECONFIG env var
- Using octant with --kubeconfig param will work regardless of the KUBECONFIG current values

**What actually happened?**

The behaviour isn't consistent and I couldn't pin-point the exact time when this happens. See my try-and-error below (i renamed the yamls for readability)

```console
# setting KUBECONFIG to a single file, cluster-a
─ KUBECONFIG=/Users/omerbarel/.kube/cluster-a.yaml

# running octant works as expected
─ octant
2020-01-06T10:33:59.303+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/serviceEditor", "module-name": "overview"}
2020-01-06T10:33:59.309+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "deployment/configuration", "module-name": "overview"}
2020-01-06T10:33:59.309+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/containerEditor", "module-name": "overview"}
2020-01-06T10:33:59.309+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "octant/deleteObject", "module-name": "configuration"}
2020-01-06T10:33:59.312+0200	INFO	dash/dash.go:370	Using embedded Octant frontend
2020-01-06T10:33:59.324+0200	INFO	dash/dash.go:349	Dashboard is available at http://127.0.0.1:7777

^C2020-01-06T10:34:25.045+0200	INFO	cluster/cluster.go:137	removing cluster client temporary directory	{"component": "cluster client"}
2020-01-06T10:34:25.046+0200	INFO	plugin/manager.go:317	shutting down plugin watcher

# running octant with the same file cluster-a via the --kubeconfig param also works:

─ octant --kubeconfig ~/.kube/cluster-a.yaml
2020-01-06T10:34:43.700+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "deployment/configuration", "module-name": "overview"}
2020-01-06T10:34:43.701+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/containerEditor", "module-name": "overview"}
2020-01-06T10:34:43.701+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/serviceEditor", "module-name": "overview"}
2020-01-06T10:34:43.701+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "octant/deleteObject", "module-name": "configuration"}
2020-01-06T10:34:43.701+0200	INFO	dash/dash.go:370	Using embedded Octant frontend
2020-01-06T10:34:43.714+0200	INFO	dash/dash.go:349	Dashboard is available at http://127.0.0.1:7777

^C2020-01-06T10:34:55.478+0200	INFO	cluster/cluster.go:137	removing cluster client temporary directory	{"component": "cluster client"}

# setting KUBECONFIG to a single file, cluster-b

─ KUBECONFIG=/Users/omerbarel/.kube/cluster-b.yaml

# running octant works as expected

─ octant
2020-01-06T10:35:17.162+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/serviceEditor", "module-name": "overview"}
2020-01-06T10:35:17.162+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "deployment/configuration", "module-name": "overview"}
2020-01-06T10:35:17.162+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/containerEditor", "module-name": "overview"}
2020-01-06T10:35:17.162+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "octant/deleteObject", "module-name": "configuration"}
2020-01-06T10:35:17.165+0200	INFO	dash/dash.go:370	Using embedded Octant frontend
2020-01-06T10:35:17.173+0200	INFO	dash/dash.go:349	Dashboard is available at http://127.0.0.1:7777

^C2020-01-06T10:35:31.973+0200	INFO	plugin/manager.go:317	shutting down plugin watcher
2020-01-06T10:35:31.973+0200	INFO	cluster/cluster.go:137	removing cluster client temporary directory	{"component": "cluster client"}

# running octant with the same file cluster-b via the --kubeconfig param also works:

─ octant --kubeconfig ~/.kube/cluster-b.yaml
2020-01-06T10:35:52.655+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/serviceEditor", "module-name": "overview"}
2020-01-06T10:35:52.656+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "deployment/configuration", "module-name": "overview"}
2020-01-06T10:35:52.656+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/containerEditor", "module-name": "overview"}
2020-01-06T10:35:52.656+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "octant/deleteObject", "module-name": "configuration"}
2020-01-06T10:35:52.656+0200	INFO	dash/dash.go:370	Using embedded Octant frontend
2020-01-06T10:35:52.665+0200	INFO	dash/dash.go:349	Dashboard is available at http://127.0.0.1:7777

^C2020-01-06T10:36:04.753+0200	INFO	cluster/cluster.go:137	removing cluster client temporary directory	{"component": "cluster client"}

# KUBECONFIG is still set to cluster-b
# passing cluster-a as parameter value, also works:

─ octant --kubeconfig ~/.kube/cluster-a.yaml
2020-01-06T10:36:18.340+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "deployment/configuration", "module-name": "overview"}
2020-01-06T10:36:18.341+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/containerEditor", "module-name": "overview"}
2020-01-06T10:36:18.341+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/serviceEditor", "module-name": "overview"}
2020-01-06T10:36:18.341+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "octant/deleteObject", "module-name": "configuration"}
2020-01-06T10:36:18.342+0200	INFO	dash/dash.go:370	Using embedded Octant frontend
2020-01-06T10:36:18.349+0200	INFO	dash/dash.go:349	Dashboard is available at http://127.0.0.1:7777

^C2020-01-06T10:36:26.929+0200	INFO	cluster/cluster.go:137	removing cluster client temporary directory	{"component": "cluster client"}
2020-01-06T10:36:26.930+0200	INFO	plugin/manager.go:317	shutting down plugin watcher

# creating a merged KUBECONFIG env:

KUBECONFIG=/Users/omerbarel/.kube/cluster-b.yaml:/Users/omerbarel/.kube/cluster-a.yaml

# running octant still works, cluster b was selected automatically:

─ kubectx
cluster-a
*cluster-b*

─ octant
2020-01-06T10:37:01.740+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "deployment/configuration", "module-name": "overview"}
2020-01-06T10:37:01.741+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/containerEditor", "module-name": "overview"}
2020-01-06T10:37:01.741+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/serviceEditor", "module-name": "overview"}
2020-01-06T10:37:01.741+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "octant/deleteObject", "module-name": "configuration"}
2020-01-06T10:37:01.743+0200	INFO	dash/dash.go:370	Using embedded Octant frontend
2020-01-06T10:37:01.752+0200	INFO	dash/dash.go:349	Dashboard is available at http://127.0.0.1:7777

^C2020-01-06T10:37:07.408+0200	INFO	cluster/cluster.go:137	removing cluster client temporary directory	{"component": "cluster client"}
2020-01-06T10:37:07.408+0200	INFO	plugin/manager.go:317	shutting down plugin watcher

# switching context

─ kubectx cluster-a
Switched to context "cluster-a.

# running octant fails

─ octant
2020-01-06T10:37:15.157+0200	ERROR	commands/dash.go:96	dashboard failed	{"err": "failed to init cluster client: invalid configuration: [context was not found for specified context: cluster-a, cluster has no server defined]"}
github.com/vmware-tanzu/octant/internal/commands.newOctantCmd.func1.2
	/private/tmp/octant-20191108-95744-1plrk02/src/github.com/vmware-tanzu/octant/internal/commands/dash.go:96
```

I'm on a Mac with 2 kubeconfigs, the current-context is in the 2nd file in $KUBECONFIG.
Octant crashes a second after showing the visual prompt for kubeconfig in the browser.

```
$ octant version
Version:  0.14.1
Git commit:  51ce7f6a7e87c9cfcabc9bf402dff7eda5d6986b
Built:  2020-07-20T23:52:37Z

$ echo $KUBECONFIG
/Users/jldec/.kube/config:/Users/jldec/tss/.kubeconfig

$ kubectl config get-contexts
CURRENT   NAME                 CLUSTER              AUTHINFO                  NAMESPACE
*         decc                 decc-scdc1-staging   scdc1-staging-tss-admin   scdc1-staging-tss
          docker-desktop       docker-desktop       docker-desktop            
          docker-for-desktop   docker-desktop       docker-desktop            

$ octant
2020-07-24T10:01:16.801+0100	INFO	dash/dash.go:105	no valid kube config found, initializing loading API
2020-07-24T10:01:16.801+0100	INFO	dash/dash.go:140	waiting for kube config ...
2020-07-24T10:01:16.805+0100	INFO	dash/dash.go:497	Dashboard is available at http://127.0.0.1:7777

2020-07-24T10:01:18.614+0100	ERROR	dash/dash.go:557	cannot create api: failed to init cluster client, does your kube config have a current-context set?: invalid configuration: [context was not found for specified context: decc, cluster has no server defined]
github.com/vmware-tanzu/octant/pkg/dash.(*Runner).startAPIService
	/__w/octant/octant/pkg/dash/dash.go:557
github.com/vmware-tanzu/octant/pkg/dash.(*Runner).Start.func2
	/__w/octant/octant/pkg/dash/dash.go:146
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x23750c6]

goroutine 39 [running]:
github.com/vmware-tanzu/octant/internal/api.(*API).Handler(0x0, 0x49dbe80, 0xc00019e8d0, 0xc0, 0x278fe60, 0x27b7f01, 0xc0001d9200)
	/__w/octant/octant/internal/api/api.go:142 +0x26
github.com/vmware-tanzu/octant/pkg/octant.(*HandlerFactory).Handler(0xc00154b410, 0x49dbe80, 0xc00019e8d0, 0x4789d68, 0xc00154f620, 0x1, 0xc00005800b)
	/__w/octant/octant/pkg/octant/handler.go:58 +0x98
github.com/vmware-tanzu/octant/pkg/dash.(*Runner).startAPIService(0xc00019ebd0, 0x49dbe80, 0xc00019e8d0, 0x49efc20, 0xc000784308, 0x0, 0xc00005800b, 0x1d, 0x0, 0x0, ...)
	/__w/octant/octant/pkg/dash/dash.go:568 +0x1cb
github.com/vmware-tanzu/octant/pkg/dash.(*Runner).Start.func2(0xc00019ebd0, 0x49efc20, 0xc000784308, 0xc000158000, 0xc000150000, 0x49dbe80, 0xc00019e8d0)
	/__w/octant/octant/pkg/dash/dash.go:146 +0x1c2
created by github.com/vmware-tanzu/octant/pkg/dash.(*Runner).Start
	/__w/octant/octant/pkg/dash/dash.go:137 +0x379
```

--- a/changelogs/unreleased/1176-GuessWhoSamFoo
+++ b/changelogs/unreleased/1176-GuessWhoSamFoo
@@ -1,0 +1,1 @@
+Fixed bug where kubeconfigs specified by multiple file paths were not watched

--- a/go.mod
+++ b/go.mod
@@ -45,6 +45,7 @@ require (
 	golang.org/x/tools v0.0.0-20200716134326-a8f9df4c9543
 	google.golang.org/grpc v1.30.0
 	google.golang.org/grpc/examples v0.0.0-20200707005602-4258d12073b4 // indirect
+	gotest.tools v2.2.0+incompatible // indirect
 	k8s.io/api v0.19.0-alpha.3
 	k8s.io/apiextensions-apiserver v0.19.0-alpha.3
 	k8s.io/apimachinery v0.19.0-beta.2

--- a/pkg/dash/dash.go
+++ b/pkg/dash/dash.go
@@ -97,7 +97,7 @@ func NewRunner(ctx context.Context, logger log.Logger, options Options) (*Runner
 	var apiService api.Service
 	var apiErr error
 
-	if err := findKubeConfig(logger, options.KubeConfig); err == nil {
+	if options.KubeConfig, err = validateKubeConfig(logger, options.KubeConfig); err == nil {
 		apiService, pluginService, apiErr = r.initAPI(ctx, logger, options)
 		if apiErr != nil {
 			return nil, fmt.Errorf("failed to start service api: %w", apiErr)
@@ -137,7 +137,8 @@ func (r *Runner) Start(ctx context.Context, logger log.Logger, options Options, 
 	if !r.apiCreated {
 		go func() {
 			if r.dash != nil {
-				if err := findKubeConfig(logger, options.KubeConfig); err != nil {
+				var err error
+				if options.KubeConfig, err = validateKubeConfig(logger, options.KubeConfig); err != nil {
 					logger.Infof("waiting for kube config ...")
 					options.KubeConfig = <-kubeConfigPath
 				}
@@ -543,24 +544,23 @@ func enableOpenCensus() error {
 	return nil
 }
 
-// findKubeConfig looks for kube config from .kube or provided by user
-func findKubeConfig(logger log.Logger, kubeConfig string) error {
-	found := false
+// validateKubeConfig returns a valid file list of kube config(s)
+func validateKubeConfig(logger log.Logger, kubeConfig string) (string, error) {
+	fileList := []string{}
 	paths := filepath.SplitList(kubeConfig)
 
 	for _, path := range paths {
 		if _, err := os.Stat(path); err == nil {
-			logger.Infof("found kube config: %v", path)
-			found = true
+			fileList = append(fileList, path)
 			continue
 		}
 		logger.Infof("cannot find kube config: %v", path)
 	}
 
-	if found {
-		return nil
+	if len(fileList) > 0 {
+		return strings.Join(fileList, ":"), nil
 	}
-	return fmt.Errorf("no kubeconfig found")
+	return "", fmt.Errorf("no kubeconfig found")
 }
 
 func (r *Runner) startAPIService(ctx context.Context, logger log.Logger, options Options) {

--- a/pkg/dash/dash.go
+++ b/pkg/dash/dash.go
@@ -19,6 +19,7 @@ import (
 	"contrib.go.opencensus.io/exporter/jaeger"
 	"github.com/skratchdot/open-golang/open"
 	"github.com/soheilhy/cmux"
+	"github.com/spf13/afero"
 	"github.com/spf13/viper"
 	"go.opencensus.io/trace"
 
@@ -69,6 +70,7 @@ type Runner struct {
 	actionManager          *action.Manager
 	websocketClientManager *api.WebsocketClientManager
 	apiCreated             bool
+	fs                     afero.Fs
 }
 
 func NewRunner(ctx context.Context, logger log.Logger, options Options) (*Runner, error) {
@@ -97,7 +99,9 @@ func NewRunner(ctx context.Context, logger log.Logger, options Options) (*Runner
 	var apiService api.Service
 	var apiErr error
 
-	if options.KubeConfig, err = validateKubeConfig(logger, options.KubeConfig); err == nil {
+	r.fs = afero.NewOsFs()
+
+	if options.KubeConfig, err = ValidateKubeConfig(logger, options.KubeConfig, r.fs); err == nil {
 		apiService, pluginService, apiErr = r.initAPI(ctx, logger, options)
 		if apiErr != nil {
 			return nil, fmt.Errorf("failed to start service api: %w", apiErr)
@@ -138,7 +142,7 @@ func (r *Runner) Start(ctx context.Context, logger log.Logger, options Options, 
 		go func() {
 			if r.dash != nil {
 				var err error
-				if options.KubeConfig, err = validateKubeConfig(logger, options.KubeConfig); err != nil {
+				if options.KubeConfig, err = ValidateKubeConfig(logger, options.KubeConfig, r.fs); err != nil {
 					logger.Infof("waiting for kube config ...")
 					options.KubeConfig = <-kubeConfigPath
 				}
@@ -544,13 +548,18 @@ func enableOpenCensus() error {
 	return nil
 }
 
-// validateKubeConfig returns a valid file list of kube config(s)
-func validateKubeConfig(logger log.Logger, kubeConfig string) (string, error) {
+// ValidateKubeConfig returns a valid file list of kube config(s)
+func ValidateKubeConfig(logger log.Logger, kubeConfig string, fs afero.Fs) (string, error) {
 	fileList := []string{}
 	paths := filepath.SplitList(kubeConfig)
 
 	for _, path := range paths {
-		if _, err := os.Stat(path); err == nil {
+		exists, err := afero.Exists(fs, path)
+		if err != nil {
+			logger.Errorf("check path exists: %v", err)
+		}
+
+		if exists {
 			fileList = append(fileList, path)
 			continue
 		}
@@ -558,7 +567,7 @@ func validateKubeConfig(logger log.Logger, kubeConfig string) (string, error) {
 	}
 
 	if len(fileList) > 0 {
-		return strings.Join(fileList, ":"), nil
+		return strings.Join(fileList, string(filepath.ListSeparator)), nil
 	}
 	return "", fmt.Errorf("no kubeconfig found")
 }

--- a/pkg/dash/dash.go
+++ b/pkg/dash/dash.go
@@ -12,6 +12,7 @@ import (
 	"net"
 	"net/http"
 	"os"
+	"path/filepath"
 	"strings"
 	"time"
 
@@ -544,8 +545,19 @@ func enableOpenCensus() error {
 
 // findKubeConfig looks for kube config from .kube or provided by user
 func findKubeConfig(logger log.Logger, kubeConfig string) error {
-	if _, err := os.Stat(kubeConfig); err == nil {
-		logger.Infof("using kube config: %v", kubeConfig)
+	found := false
+	paths := filepath.SplitList(kubeConfig)
+
+	for _, path := range paths {
+		if _, err := os.Stat(path); err == nil {
+			logger.Infof("found kube config: %v", path)
+			found = true
+			continue
+		}
+		logger.Infof("cannot find kube config: %v", path)
+	}
+
+	if found {
 		return nil
 	}
 	return fmt.Errorf("no kubeconfig found")

--- a/pkg/dash/dash_test.go
+++ b/pkg/dash/dash_test.go
@@ -1,0 +1,83 @@
+/*
+ * Copyright (c) 2020 the Octant contributors. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+package dash
+
+import (
+	"path/filepath"
+	"testing"
+
+	"github.com/spf13/afero"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/vmware-tanzu/octant/internal/log"
+)
+
+func TestRunner_ValidateKubeconfig(t *testing.T) {
+	fs := afero.NewMemMapFs()
+	afero.WriteFile(fs, "/test1", []byte(""), 0755)
+	afero.WriteFile(fs, "/test2", []byte(""), 0755)
+
+	separator := string(filepath.ListSeparator)
+
+	tests := []struct {
+		name     string
+		fileList string
+		expected string
+		isErr    bool
+	}{
+		{
+			name:     "single path",
+			fileList: "/test1",
+			expected: "/test1",
+			isErr:    false,
+		},
+		{
+			name:     "multiple paths",
+			fileList: "/test1" + separator + "/test2",
+			expected: "/test1" + separator + "/test2",
+			isErr:    false,
+		},
+		{
+			name:     "single path not found",
+			fileList: "/unknown",
+			expected: "",
+			isErr:    true,
+		},
+		{
+			name:     "multiple paths not found",
+			fileList: "/unknown" + separator + "/unknown2",
+			expected: "",
+			isErr:    true,
+		},
+		{
+			name:     "multiple file path; missing a config",
+			fileList: "/test1" + separator + "/unknown",
+			expected: "/test1",
+			isErr:    false,
+		},
+		{
+			name:     "invalid path",
+			fileList: "not a filepath",
+			expected: "",
+			isErr:    true,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			logger := log.NopLogger()
+			path, err := ValidateKubeConfig(logger, test.fileList, fs)
+			if test.isErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+			assert.Equal(t, path, test.expected)
+		})
+	}
+}

